### DISCUSSION TITLE
Should Django template context/parent variables be automatically passed down or need to be explictily passed?

### DISCUSSION THREAD

**Discussion by mattbha** [wrabit/django-cotton#156] [General]
![theory](https://github.com/user-attachments/assets/82469d08-4f4e-492b-9ad7-ea4164671d24)

Lets say I have an example structure like the above.

It appears to work fine without the `:event=event` - which I'm not sure is the best behavior. 

My question for debate is, should this be required (to pass the event object) down to the component? Or should the for loop implicitly send it to the component. I can see an argument for both ways. Currently, it appears the parameter `:event=event` does not need to be specified, but then I am left wondering what parameters are being sent to the component.

Having it automatically passed down decreases verbosity (a win), but does leave me wondering/tracing back to where the object may come from, which can be complex. Especially if its sent from a django views.py render context. In this case, `{{ event }}` may not even be present in the parent template (say `view_event.html` but it would be accessible in the component without the explicit `:event=event`.

I'd personally vote to err on the side of the zen of python 

> Explicit is better than implicit.

So I am trying to include (the currently non-required) code of `:event=event` everywhere. (but for example I am not passing `:request=request`, which makes the design feel a bit odd.) Part of me thinks limiting the component scope to whats passed in rather than inherited from the parent context makes better sense, since the scopes are typically across files.

What does everyone think?

**Comment by mattbha** [comment]
A real example. Here is a template from my project. `{{ event }}` is not even used in this parent template, but the `event.event-header` component does get the event, with or without `:event=event` being removed. It just feels a bit funny to me.

```
{% extends 'backend/base.html' %}

{% block content %}
<div class="max-w-full mx-auto">
    <c-event.event-header :event=event :editable=True />
    ...
</div>
{% endblock content %}
```

**Comment by wrabit** [comment]
@mattbha This behaviour actually is unintentional and came about since the v1.0.x rewrite. Previous versions would isolate the context and I think I much prefer that. It caught me out just this evening where a variable had 'leaked' into the component. I think I will work on re-instating this functionality in the next release so hold on for this update!

**Reply by wrabit** [reply]
@mattbha v1.0.7 released which mirrors django `only` / isolated context in components.

**Reply by wrabit** [reply]
@mattbha this was not the behaviour prior to v1 and I don't know why I thought that. I think because crossing over from js land, I just defined all incoming props by default. The issue is that, isolating context  removes django builtin context variables like `csrf_token`, `user`, `request`, `perms`, `messages` additionally any custom vars from context processors you may have in your app. 

I think forcing these as props is anti-productive:

`<c-user-form request=request csrf_token=csrf_token custom_ctx_var=custom_context_var />`

One option is to manage this on cotton side to explicitly include those common items along with the component's isolated context. But only if we can determine and include all 'default' items and without greatly affecting performance. 

I will take a look but if that doesn't get us anywhere I think the only sensible thing is to revert the recent change for now.

**Reply by mattbha** [reply]
well then it makes sense to allow it, and then use `only` to get the isolated context (similar to django) 

similar to https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#include
